frontend/dockerfile/dockerfile2llb/convert.go

```diff
@@ -79,22 +87,75 @@ type ConvertOpt struct {
 }
 
 func Dockerfile2LLB(ctx context.Context, dt []byte, opt ConvertOpt) (*llb.State, *Image, *binfotypes.BuildInfo, error) {
-	ds, bi, err := toDispatchState(ctx, dt, opt)
+	ds, _, bi, err := toDispatchState(ctx, dt, opt)
 	if err != nil {
 		return nil, nil, nil, err
 	}
 	return &ds.state, &ds.image, bi, nil
 }
 
 func Dockefile2Outline(ctx context.Context, dt []byte, opt ConvertOpt) (*outline.Outline, error) {
-	ds, _, err := toDispatchState(ctx, dt, opt)
+	ds, _, _, err := toDispatchState(ctx, dt, opt)
 	if err != nil {
 		return nil, err
 	}
 	o := ds.Outline(dt)
 	return &o, nil
 }
 
+type SBOMTargets struct {
+	Core   llb.State
+	Extras map[string]llb.State
+}
+
+func Dockerfile2SBOMTargets(ctx context.Context, dt []byte, opt ConvertOpt) (*SBOMTargets, error) {
+	ds, dss, _, err := toDispatchState(ctx, dt, opt)
```

Comment:
I would like this a lot more if we would not call `toDispatchState` twice in here. Calling this twice means that LLB is generated twice, images are resolved twice, callbacks in `opt` all called twice etc. There are so many things that could go wrong here as none of the code is meant to handle this case. Even LLB could generate completely different state as in some cases uniqueness parameters are used in locals. The other cases are probably even messier.

Better to build up all the state you need for sboms during `toDispatchState` to the `dispatchState` properties , and then return it out in `Dockerfile2LLB` as separate parameter.